src/cli.ts

```diff
@@ -25,35 +39,135 @@ async function main(argv: minimist.ParsedArgs) {
-    await writeOutput(ts, argOut)
+    // Process input as either glob, directory, or single file
+    if (ISGLOB) {
+      await processGlob(argIn, argOut, argv as Partial<Options>)
+    } else if (ISDIR) {
+      await processDir(argIn, argOut, argv as Partial<Options>)
+    } else {
+      await processFile(argIn, argOut, argv as Partial<Options>)
+    }
   } catch (e) {
     console.error(whiteBright.bgRedBright('error'), e)
     process.exit(1)
   }
 }
 
-function readInput(argIn?: string) {
-  if (!argIn) {
-    return new Promise(stdin)
+// check if path is an existing directory
+function isDir(path: string): boolean {
+  return existsSync(path) && lstatSync(path).isDirectory()
+}
+
+async function processGlob(argIn: string, argOut: string | undefined, argv: Partial<Options>) {
+  const files = await glob(argIn) // execute glob pattern match
+
+  if (files.length === 0) {
+    throw ReferenceError(
+      `You passed a glob pattern "${argIn}", but there are no files that match that pattern in ${process.cwd()}`
+    )
   }
-  return readFile(resolve(process.cwd(), argIn), 'utf-8')
+  // create output directory if it does not exist
+  if (argOut && !existsSync(argOut)) {
+    await mkdirp(argOut)
+  }
+
+  Promise.all(
+    files.map(file => {
+      const outPath = argOut && `${argOut}/${basename(file, '.json')}.d.ts`
+      processFile(file, outPath, argv)
+    })
+  )
+}
+
+/*
+the following logic determines the out path by comparing the in path to the users specified out path.
+For example, if input directory MultiSchema looks like:
+  MultiSchema/foo/a.json
+  MultiSchema/bar/fuzz/c.json
+  MultiSchema/bar/d.json
+And the user wants the outputs to be in MultiSchema/Out, then this code will be able to map the inner directories foo, bar, and fuzz into the intended Out directory like so:
+  MultiSchema/Out/foo/a.json
+  MultiSchema/Out/bar/fuzz/c.json
+  MultiSchema/Out/bar/d.json
+*/
+function pathTransform(o: string, i: string): string {
+  const outPathList = o.split('/')
+  const inPathList = i.split('/')
+
+  const intersection = outPathList.filter(x => inPathList.includes(x))
+  const difference = outPathList
+    .filter(x => !inPathList.includes(x))
+    .concat(inPathList.filter(x => !outPathList.includes(x)))
+
+  return join(...intersection, ...difference)
+}
+
+async function processDir(argIn: string, argOut: string | undefined, argv: Partial<Options>): Promise<void[]> {
+  const files = getPaths(argIn)
+
+  return Promise.all(
+    files.map(file => {
+      if (!argOut) {
+        processFile(file, argOut, argv)
+      } else {
+        let outPath = pathTransform(argOut, file)
+        // const outPathList = argOut.split('/')
+        // const inPathList = file.split('/')
+
+        // const intersection = outPathList.filter(x => inPathList.includes(x))
+        // const difference = outPathList
+        //   .filter(x => !inPathList.includes(x))
+        //   .concat(inPathList.filter(x => !outPathList.includes(x)))
```

Comment:
Are these intentionally commented out?

Reply:
Ahh oops meant to delete after i put it into its own function. Fixing now (and adding tests 😉)